File: src/crypto/signers/private-key-signer.ts
/**
 * WASM implementation of PrivateKeySigner
 * Provides wallet/signer functionality using WASM Zig primitives
 */

import { secp256k1 } from "@noble/curves/secp256k1.js";
import * as BrandedAddress from "../../primitives/Address/BrandedAddress/index.js";
import { Address } from "../../primitives/Address/index.js";
import * as primitives from "../../wasm-loader/loader.js";
import { Keccak256Wasm } from "../keccak256.wasm.js";

/**
 * Options for creating a private key signer.
 *
 * @see https://voltaire.tevm.sh/crypto for crypto documentation
 * @since 0.0.0
 */
export interface PrivateKeySignerOptions {
	/** Private key as hex string (with or without 0x prefix) or Uint8Array (32 bytes) */
	privateKey: string | Uint8Array;
}

/**
 * Interface for Ethereum signer that can sign messages, transactions, and typed data.
 *
 * @see https://voltaire.tevm.sh/crypto for crypto documentation
 * @since 0.0.0
 */
export interface Signer {
	/** The checksummed Ethereum address */
	address: string;
	/** The uncompressed public key (64 bytes, without 0x04 prefix) */
	publicKey: Uint8Array;
	/** Sign a message with EIP-191 prefix */
	signMessage(message: string | Uint8Array): Promise<string>;
	/** Sign a transaction */
	signTransaction(transaction: any): Promise<any>;
	/** Sign typed data according to EIP-712 */
	signTypedData(typedData: any): Promise<string>;
}

/**
 * WASM-based implementation of Ethereum private key signer.
 *
 * Uses Zig primitives via WASM for cryptographic operations. The private key is
 * securely stored in a closure and never exposed publicly.
 *
 * @see https://voltaire.tevm.sh/crypto for crypto documentation
 * @since 0.0.0
 * @example
 * ```javascript
 * import { PrivateKeySignerImpl } from './crypto/signers/private-key-signer.js';
 *
 * // Create from hex string
 * const signer = PrivateKeySignerImpl.fromPrivateKey({
 *   privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
 * });
 *
 * // Sign a message
 * const signature = await signer.signMessage('Hello, Ethereum!');
 *
 * // Sign a transaction
 * const signedTx = await signer.signTransaction({
 *   to: '0x...',
 *   value: 1000000000000000000n,
 *   // ... other tx fields
 * });
 * ```
 */
export class PrivateKeySignerImpl implements Signer {
	public readonly address: string;
	public readonly publicKey: Uint8Array;
	private readonly signWithPrivateKey: (hash: Uint8Array) => Uint8Array;

	private constructor(privateKey: Uint8Array) {
		// Private key only exists in this closure
		this.signWithPrivateKey = (hash: Uint8Array) => {
			const sig = primitives.secp256k1Sign(hash, privateKey);
			// Return 65-byte signature: r[32] + s[32] + v[1]
			const result = new Uint8Array(65);
			result.set(sig.r, 0);
			result.set(sig.s, 32);
			result[64] = sig.v;
			return result;
		};

		// Derive public key from private key
		const pubKey = secp256k1.getPublicKey(privateKey, false);
		this.publicKey = pubKey.slice(1); // Remove 0x04 prefix for uncompressed

		// Derive address from public key (keccak256(pubkey)[12:])
		const pubkeyHash = Keccak256Wasm.hash(this.publicKey);
		const addressBytes = pubkeyHash.slice(-20);
		const addressObj = Address.fromBytes(addressBytes);
		this.address = BrandedAddress.toChecksummed(addressObj);
	}

	/**
	 * Creates a private key signer from a hex string or Uint8Array.
	 *
	 * @see https://voltaire.tevm.sh/crypto for crypto documentation
	 * @since 0.0.0
	 * @param {PrivateKeySignerOptions} options - Options containing the private key
	 * @returns {PrivateKeySignerImpl} A new signer instance
	 * @throws {Error} If private key is not 32 bytes or has invalid format
	 * @example
	 * ```javascript
	 * import { PrivateKeySignerImpl } from './crypto/signers/private-key-signer.js';
	 *
	 * // From hex string with 0x prefix
	 * const signer1 = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
	 * });
	 *
	 * // From hex string without 0x prefix
	 * const signer2 = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: 'ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
	 * });
	 *
	 * // From Uint8Array
	 * const privateKeyBytes = new Uint8Array(32);
	 * const signer3 = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: privateKeyBytes
	 * });
	 * ```
	 */
	static fromPrivateKey(
		options: PrivateKeySignerOptions,
	): PrivateKeySignerImpl {
		let privateKeyBytes: Uint8Array;

		if (typeof options.privateKey === "string") {
			const hex = options.privateKey.startsWith("0x")
				? options.privateKey.slice(2)
				: options.privateKey;
			if (hex.length !== 64) {
				throw new Error("Private key must be 32 bytes (64 hex characters)");
			}
			privateKeyBytes = new Uint8Array(
				hex.match(/.{1,2}/g)?.map((byte) => Number.parseInt(byte, 16)) ?? [],
			);
		} else {
			privateKeyBytes = new Uint8Array(options.privateKey);
		}

		if (privateKeyBytes.length !== 32) {
			throw new Error("Private key must be 32 bytes");
		}

		return new PrivateKeySignerImpl(privateKeyBytes);
	}

	/**
	 * Signs a message using EIP-191 personal sign format.
	 *
	 * Adds the Ethereum signed message prefix before hashing and signing.
	 *
	 * @see https://voltaire.tevm.sh/crypto for crypto documentation
	 * @since 0.0.0
	 * @param {string | Uint8Array} message - The message to sign
	 * @returns {Promise<string>} The signature as a hex string (0x + r + s + v, 65 bytes)
	 * @throws {never}
	 * @example
	 * ```javascript
	 * import { PrivateKeySignerImpl } from './crypto/signers/private-key-signer.js';
	 *
	 * const signer = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
	 * });
	 *
	 * // Sign a string message
	 * const sig1 = await signer.signMessage('Hello, Ethereum!');
	 * console.log(sig1); // '0x...' (130 hex chars)
	 *
	 * // Sign a Uint8Array message
	 * const msgBytes = new TextEncoder().encode('Hello');
	 * const sig2 = await signer.signMessage(msgBytes);
	 * ```
	 */
	async signMessage(message: string | Uint8Array): Promise<string> {
		// Hash message with EIP-191 prefix
		const msgBytes =
			typeof message === "string" ? new TextEncoder().encode(message) : message;
		const prefix = new TextEncoder().encode(
			`\x19Ethereum Signed Message:\n${msgBytes.length}`,
		);
		const combined = new Uint8Array(prefix.length + msgBytes.length);
		combined.set(prefix);
		combined.set(msgBytes, prefix.length);
		const messageHash = Keccak256Wasm.hash(combined);

		// Sign the hash
		const signature: Uint8Array = this.signWithPrivateKey(messageHash);
		return `0x${Array.from(signature)
			.map((b) => b.toString(16).padStart(2, "0"))
			.join("")}`;
	}

	/**
	 * Signs an Ethereum transaction.
	 *
	 * Supports all transaction types: Legacy (0), EIP-2930 (1), EIP-1559 (2), EIP-4844 (3), and EIP-7702 (4).
	 * For legacy transactions, the v value includes the chain ID. For modern transactions, yParity is used.
	 *
	 * @see https://voltaire.tevm.sh/crypto for crypto documentation
	 * @since 0.0.0
	 * @param {any} transaction - The transaction object to sign
	 * @returns {Promise<any>} The signed transaction with signature fields (r, s, v/yParity)
	 * @throws {never}
	 * @example
	 * ```javascript
	 * import { PrivateKeySignerImpl } from './crypto/signers/private-key-signer.js';
	 *
	 * const signer = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
	 * });
	 *
	 * // Sign an EIP-1559 transaction
	 * const signedTx = await signer.signTransaction({
	 *   type: 2,
	 *   chainId: 1n,
	 *   nonce: 0n,
	 *   maxFeePerGas: 20000000000n,
	 *   maxPriorityFeePerGas: 1000000000n,
	 *   gas: 21000n,
	 *   to: '0x...',
	 *   value: 1000000000000000000n,
	 *   data: '0x'
	 * });
	 * console.log(signedTx.yParity, signedTx.r, signedTx.s);
	 * ```
	 */
	async signTransaction(transaction: any): Promise<any> {
		const Transaction = await import(
			"../../primitives/Transaction/index.js"
		).then((m) => m);

		// Get signing hash
		const signingHash = Transaction.getSigningHash(transaction);

		// Sign the hash
		const signature: Uint8Array = this.signWithPrivateKey(signingHash);

		// Parse signature (65 bytes: r[32] + s[32] + v[1])
		const r = signature.slice(0, 32);
		const s = signature.slice(32, 64);
		const v = signature[64] ?? 0;

		// Convert v from 0-1 to appropriate format based on tx type
		let yParity: number;
		if (transaction.type === 0) {
			// Legacy: v includes chain ID
			const chainId = transaction.chainId ?? 0n;
			const vValue = BigInt(v) + 35n + chainId * 2n;
			return { ...transaction, v: vValue, r, s };
		}
		// EIP-2930, EIP-1559, EIP-4844, EIP-7702: use yParity
		yParity = v;

		return { ...transaction, yParity, r, s };
	}

	/**
	 * Signs typed structured data according to EIP-712.
	 *
	 * Hashes the typed data using the EIP-712 standard and signs the hash.
	 *
	 * @see https://voltaire.tevm.sh/crypto for crypto documentation
	 * @since 0.0.0
	 * @param {any} typedData - The EIP-712 typed data object
	 * @returns {Promise<string>} The signature as a hex string (0x + r + s + v, 65 bytes)
	 * @throws {never}
	 * @example
	 * ```javascript
	 * import { PrivateKeySignerImpl } from './crypto/signers/private-key-signer.js';
	 *
	 * const signer = PrivateKeySignerImpl.fromPrivateKey({
	 *   privateKey: '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80'
	 * });
	 *
	 * const typedData = {
	 *   types: {
	 *     EIP712Domain: [
	 *       { name: 'name', type: 'string' },
	 *       { name: 'version', type: 'string' }
	 *     ],
	 *     Message: [{ name: 'content', type: 'string' }]
	 *   },
	 *   primaryType: 'Message',
	 *   domain: { name: 'MyApp', version: '1' },
	 *   message: { content: 'Hello' }
	 * };
	 *
	 * const signature = await signer.signTypedData(typedData);
	 * console.log(signature); // '0x...'
	 * ```
	 */
	async signTypedData(typedData: any): Promise<string> {
		const { Eip712Wasm } = await import("../eip712.wasm.js");

		// Hash typed data according to EIP-712
		const hash = Eip712Wasm.hashTypedData(typedData);

		// Sign the hash
		const signature: Uint8Array = this.signWithPrivateKey(hash);

		// Return as hex string (0x + r + s + v)
		return `0x${Array.from(signature)
			.map((b) => b.toString(16).padStart(2, "0"))
			.join("")}`;
	}
}

export default PrivateKeySignerImpl;

File: src/crypto/signers/private_key_signer.go
package signers

import (
	"crypto/rand"
	"errors"
	"fmt"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/decred/dcrd/dcrec/secp256k1/v4/ecdsa"
	"github.com/evmts/primitives/src/crypto"
)

var (
	ErrInvalidPrivateKey = errors.New("invalid private key")
	ErrSigningFailed     = errors.New("signing failed")
)

// PrivateKeySigner signs transactions and messages using a private key.
//
// SECURITY CRITICAL:
// - Private keys are stored in memory and should be cleared after use
// - Uses decred/secp256k1 library (well-audited, deterministic signing)
// - Implements EIP-2 (no malleable signatures)
// - Supports EIP-191 (personal message signing)
type PrivateKeySigner struct {
	privateKey *secp256k1.PrivateKey
	publicKey  *secp256k1.PublicKey
	address    [20]byte
}

// NewPrivateKeySigner creates a signer from a 32-byte private key.
//
// Parameters:
//   - privateKeyBytes: 32-byte private key
//
// Returns:
//   - *PrivateKeySigner: Initialized signer
//   - error: ErrInvalidPrivateKey if key is invalid
func NewPrivateKeySigner(privateKeyBytes [32]byte) (*PrivateKeySigner, error) {
	// Validate and create private key
	privateKey := secp256k1.PrivKeyFromBytes(privateKeyBytes[:])

	// Derive public key
	publicKey := privateKey.PubKey()

	// Derive Ethereum address
	address := deriveAddress(publicKey)

	return &PrivateKeySigner{
		privateKey: privateKey,
		publicKey:  publicKey,
		address:    address,
	}, nil
}

// Random generates a new random private key signer.
//
// Returns:
//   - *PrivateKeySigner: Initialized signer with random key
//   - error: if random number generation fails
func Random() (*PrivateKeySigner, error) {
	var privateKeyBytes [32]byte
	_, err := rand.Read(privateKeyBytes[:])
	if err != nil {
		return nil, fmt.Errorf("failed to generate random key: %w", err)
	}

	return NewPrivateKeySigner(privateKeyBytes)
}

// Address returns the Ethereum address associated with this signer.
func (s *PrivateKeySigner) Address() [20]byte {
	return s.address
}

// GetPrivateKey returns the private key bytes.
//
// SECURITY WARNING: This exposes the private key. Use with extreme caution.
// Consider using SecureZero() on the returned bytes after use.
func (s *PrivateKeySigner) GetPrivateKey() [32]byte {
	var key [32]byte
	copy(key[:], s.privateKey.Serialize())
	return key
}

// GetPublicKey returns the uncompressed public key (64 bytes: x || y).
func (s *PrivateKeySigner) GetPublicKey() [64]byte {
	var pubkey [64]byte

	// Get uncompressed public key (65 bytes: 0x04 || x || y)
	uncompressed := s.publicKey.SerializeUncompressed()

	// Skip the 0x04 prefix byte
	copy(pubkey[:], uncompressed[1:])

	return pubkey
}

// SignMessageHash signs a pre-hashed message.
//
// Parameters:
//   - hash: 32-byte message hash
//
// Returns:
//   - r, s: 32-byte signature components
//   - v: recovery ID (0 or 1)
//   - error: if signing fails
func (signer *PrivateKeySigner) SignMessageHash(hash [32]byte) (r, s [32]byte, v byte, err error) {
	// Sign the hash using RFC 6979 deterministic signing
	signature := ecdsa.Sign(signer.privateKey, hash[:])

	// Extract r and s components as fixed arrays
	sigR := signature.R()
	sigS := signature.S()
	r = sigR.Bytes()
	s = sigS.Bytes()

	// Calculate recovery ID
	// We need to try both recovery IDs and see which one recovers to our public key
	recoveryID, err := calculateRecoveryID(hash, r, s, signer.address)
	if err != nil {
		return r, s, 0, fmt.Errorf("failed to calculate recovery ID: %w", err)
	}

	v = recoveryID
	return r, s, v, nil
}

// SignMessage signs a message using EIP-191 personal message format.
//
// The message is prefixed with "\x19Ethereum Signed Message:\n{length}"
// before hashing and signing.
//
// Parameters:
//   - message: Message bytes to sign
//
// Returns:
//   - [65]byte: Signature (r || s || v)
//   - error: if signing fails
func (signer *PrivateKeySigner) SignMessage(message []byte) ([65]byte, error) {
	// Hash the message using EIP-191 format
	hash := eip191Hash(message)

	// Sign the hash
	r, s, v, err := signer.SignMessageHash(hash)
	if err != nil {
		return [65]byte{}, err
	}

	// Serialize to 65 bytes (r || s || v+27)
	// Note: EIP-191 uses v = 27 or 28 (add 27 to recovery ID)
	return crypto.SignatureToBytes(r, s, v+27), nil
}

// SignTransaction signs a transaction.
//
// Parameters:
//   - tx: Transaction to sign
//
// Returns:
//   - r, s: 32-byte signature components
//   - v: recovery ID
//   - error: if signing fails
func (signer *PrivateKeySigner) SignTransaction(tx Transaction) (r, s [32]byte, v byte, err error) {
	// Get the transaction hash to sign
	hash := tx.GetSignatureHash()

	// Sign the hash
	return signer.SignMessageHash(hash)
}

// Destroy securely zeros the private key in memory.
//
// SECURITY: Call this when the signer is no longer needed.
func (s *PrivateKeySigner) Destroy() {
	if s.privateKey != nil {
		// Zero the private key bytes
		keyBytes := s.privateKey.Serialize()
		crypto.SecureZero(keyBytes)
		s.privateKey = nil
	}
}

// deriveAddress derives an Ethereum address from a public key.
//
// Address = Keccak256(pubkey)[12:]
func deriveAddress(pubkey *secp256k1.PublicKey) [20]byte {
	// Get uncompressed public key (65 bytes: 0x04 || x || y)
	uncompressed := pubkey.SerializeUncompressed()

	// Skip the 0x04 prefix byte, hash the 64-byte x||y coordinates
	hash := crypto.Keccak256(uncompressed[1:])

	// Take last 20 bytes
	var address [20]byte
	copy(address[:], hash[12:])

	return address
}

// eip191Hash hashes a message using EIP-191 personal message format.
//
// Format: keccak256("\x19Ethereum Signed Message:\n{length}{message}")
func eip191Hash(message []byte) [32]byte {
	// Build the EIP-191 prefix
	prefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(message))

	// Concatenate prefix + message
	data := append([]byte(prefix), message...)

	// Hash with Keccak-256
	return crypto.Keccak256(data)
}

// calculateRecoveryID determines which recovery ID (0 or 1) produces the correct address.
func calculateRecoveryID(hash, r, s [32]byte, expectedAddress [20]byte) (byte, error) {
	// Try recovery ID 0
	recoveredAddr0, err0 := crypto.RecoverAddress(hash, r, s, 0)
	if err0 == nil && recoveredAddr0 == expectedAddress {
		return 0, nil
	}

	// Try recovery ID 1
	recoveredAddr1, err1 := crypto.RecoverAddress(hash, r, s, 1)
	if err1 == nil && recoveredAddr1 == expectedAddress {
		return 1, nil
	}

	// Neither recovery ID worked
	return 0, fmt.Errorf("could not determine recovery ID")
}
